File: Xamarin/Xamarin.Native/Xamarin.Native.Droid/Activities/PerformanceTestActivity.cs
using Android.App;
using Android.Content;
using Android.OS;
using Android.Runtime;
using Android.Views;
using Android.Widget;
using Java.Lang;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Xamarin.Services;

namespace Xamarin.Native.Droid.Activities
{
    [Activity(Label = "PerformanceActivity")]
    public class PerformanceTestActivity : Activity
    {
        Stopwatch stopwatch;
        Button startButton;
        EditText digitsEntry;
        TextView resultView;
        TextView timeLabel;
        ProgressDialog progressDialog;

        protected override void OnCreate(Bundle savedInstanceState)
        {
            base.OnCreate(savedInstanceState);
            SetContentView(Resource.Layout.PerformanceTest);
            Title = GetString(Resource.String.menuPerformance);

            digitsEntry = FindViewById<EditText>(Resource.Id.digitsEntry);
            resultView = FindViewById<TextView>(Resource.Id.resultView);
            timeLabel = FindViewById<TextView>(Resource.Id.timeLabel);
            startButton = FindViewById<Button>(Resource.Id.startButton);

            startButton.Click += StartCalculation;
            PerformanceTestService.Instance.PiCalculationCompleted += PiCalculationCompleted;
        }

        private void StartCalculation(object sender, EventArgs e)
        {
            stopwatch = new Stopwatch();
            stopwatch.Start();
            progressDialog = ProgressDialog.Show(this, "Przetwarzanie...", "");

            var digits = Convert.ToInt32(digitsEntry.Text);
            new Thread(new Runnable(() =>
            {
                PerformanceTestService.Instance.CalculatePi(digits);
            })).Start();
        }

        private void PiCalculationCompleted(string result)
        {
            stopwatch.Stop();
            RunOnUiThread(new Runnable(() =>
            {
                resultView.Text = result;
                timeLabel.Text = stopwatch.GetDurationInSeconds();
                progressDialog.Dismiss();
            }));
        }
    }
}

File: Native.Android/app/src/main/java/com/pgrzmil/nativedroid/PerformanceTestActivity.java
package com.pgrzmil.nativedroid;

import android.app.ProgressDialog;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import com.pgrzmil.services.*;

public class PerformanceTestActivity extends AppCompatActivity implements PerformanceTestListener {
    Stopwatch stopwatch;
    Button startButton;
    EditText digitsEntry;
    TextView timeLabel;
    TextView resultView;
    ProgressDialog progressDialog;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_performance);
        setTitle(R.string.menuPerformance);

        digitsEntry = (EditText) findViewById(R.id.digitsEntry);
        timeLabel = (TextView) findViewById(R.id.timeLabel);
        resultView = (TextView) findViewById(R.id.resultView);
        startButton = (Button) findViewById(R.id.startButton);

        stopwatch = new Stopwatch();
        PerformanceTestService.getInstance().addListener(this);
    }

    public void startCalculation(View view) {
        stopwatch.start();
        final int digits = Integer.parseInt(digitsEntry.getText().toString());
        progressDialog = ProgressDialog.show(this, "Przetwarzanie...", "");
        new Thread(new Runnable() {
            @Override
            public void run() {
                PerformanceTestService.getInstance().calculatePi(digits);
            }
        }).start();
    }

    @Override
    public void piCalculationCompleted(final String result) {
        stopwatch.stop();
        runOnUiThread(new Runnable() {
            public void run() {
                resultView.setText(result);
                timeLabel.setText(stopwatch.getDurationBreakdown());
                progressDialog.dismiss();
            }
        });
    }
}
